modules/core/src/main/java/org/apache/ignite/internal/jdbc2/JdbcBinaryBuffer.java

```diff
@@ -0,0 +1,449 @@
+     * argument {@code b}. The 24 high-order bits of {@code b}b are ignored.
+     *
+     * @param pos Pointer to a position.
+     * @param b Byte to write.
+     */
+    void write(int pos, int b) throws IOException {
+        if (MAX_ARRAY_SIZE - pos < 1)
+            throw new IOException("Too much data. Can't write more then " + MAX_ARRAY_SIZE + " bytes.");
+
+        int newLen = Math.max(pos + 1, len);
+
+        reallocateIfRequired(newLen);
+
+        arr[pos] = (byte)b;
+
+        len = newLen;
+    }
+
+    /**
+     * Ensure capacity.
+     *
+     * @param newLen The new data length the buffer should be able to hold.
+     */
+    private void reallocateIfRequired(int newLen) {
+        if (newLen - arr.length > 0 || isReadOnly)
+            reallocate(capacity(arr.length, newLen));
+    }
+
+    /**
+     * Calculate new capacity.
+     *
+     * @param cap Current capacity.
+     * @param reqCap Required new capacity.
+     * @return New capacity.
+     */
+    protected static int capacity(int cap, int reqCap) {
+        if (reqCap <= MIN_CAP)
+            return MIN_CAP;
+
+        int resCap = Math.max(cap, MIN_CAP);
+
+        while (resCap < reqCap) {
+            resCap <<= 1;
+
+            if (resCap < 0)
+                return MAX_ARRAY_SIZE;
+        }
+
+        return resCap;
+    }
+
+    /**
+     * Allocate the new underlining array and copy data.
+     *
+     * @param newCapacity New capacity.
+     */
+    private void reallocate(int newCapacity) {
+        byte[] newBuf = new byte[newCapacity];
+
+        U.arrayCopy(arr, off, newBuf, 0, len);
+
+        arr = newBuf;
+        off = 0;
+        isReadOnly = false;
+    }
+
+    /**
+     * Input stream to read data from buffer.
+     */
+    private class BufferInputStream extends InputStream {
+        /** Max position in the buffer. -1 means no max position (for unlimited stream). */
+        private final int limit;
+
+        /** Current position in the buffer. */
+        private int pos;
+
+        /** Remembered buffer position at the moment the {@link InputStream#mark} is called. */
+        private int markedPos;
+
+        /**
+         * Create unlimited stream to read all data from the buffer starting from the beginning.
+         */
+        private BufferInputStream() {
+            this(0, -1);
+        }
+
+        /**
+         * Create stream to read data from the buffer starting from the specified {@code start}
+         * zero-based position.
+         *
+         * @param start The zero-based offset to the first byte to be retrieved.
+         * @param limit The maximim length in bytes of the data to be retrieved. Unlimited if null.
+         */
+        private BufferInputStream(int start, int limit) {
+            pos = start;
+
+            markedPos = start;
+
+            this.limit = limit;
+        }
+
+        /** {@inheritDoc} */
+        @Override public int read() {
+            if (limit != -1 && pos >= limit)
```

Comment:
Looks like the same check exists in `JdbcBinaryBuffer#read`.
Can we remove this and rely to the check inside `JdbcBinaryBuffer#read`?

Reply:
Same logic as above.  Limits are imposed to stream instances, not to `JdbcBinaryBuffer`.

Reply:
We implement same logic twice - at least we can introduce some methods to perform calculation.